pkg/cmd/admin/prune/images.go

```diff
@@ -513,3 +768,312 @@ func getClients(f *clientcmd.Factory, caBundle string) (*client.Client, kclients
+	}
+
+	var result kapi.PodList = *podList
+	result.Items = []kapi.Pod{}
+	for _, pod := range podList.Items {
+		for _, label := range []struct{ key, value string }{
+			// DeploymentConfigs
+			{deployapi.DeploymentConfigLabel, "docker-registry"},
+			// DaemonSets
+			{"docker-registry", DefaultRegistryNamespace},
+			// TODO: handle upstream deployments
+		} {
+			if pod.Labels[label.key] != label.value {
+				continue
+			}
+			result.Items = append(result.Items, pod)
+			break
+		}
+	}
+
+	if len(result.Items) == 0 {
+		return nil, fmt.Errorf("no registry pods found")
+	}
+
+	return podList, nil
+}
+
+func collectRegistryDeploymentModels(
+	dcNamespacer client.DeploymentConfigsNamespacer,
+	deploymentGetter kextclientset.DeploymentsGetter,
+	dsGetter kextclientset.DaemonSetsGetter,
+) ([]runtime.Object, error) {
+	var result []runtime.Object
+
+	dcList, err := dcNamespacer.DeploymentConfigs(DefaultRegistryNamespace).List(metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{kapi.ObjectNameField: "docker-registry"}).String(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	for i := range dcList.Items {
+		dc := &dcList.Items[i]
+		result = append(result, dc)
+	}
+
+	deploymentList, err := deploymentGetter.Deployments(DefaultRegistryNamespace).List(metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{kapi.ObjectNameField: "docker-registry"}).String(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	for i := range deploymentList.Items {
+		deployment := &deploymentList.Items[i]
+		result = append(result, deployment)
+	}
+
+	dsList, err := dsGetter.DaemonSets(DefaultRegistryNamespace).List(metav1.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{kapi.ObjectNameField: "docker-registry"}).String(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	for i := range dsList.Items {
+		ds := &dsList.Items[i]
+		result = append(result, ds)
+	}
+
+	if len(result) == 0 {
+		return nil, fmt.Errorf("no deploymentconfig, daemonset nor deployment named docker-registry found in the default namespace")
+	}
+
+	return result, nil
+}
+
+func getMatchingPods(decoder runtime.Decoder, deploymentModel runtime.Object, podList *kapi.PodList) kapi.PodList {
+	result := *podList
+	result.Items = []kapi.Pod{}
+
+	for i := range podList.Items {
+		pod := &podList.Items[i]
+		sref := kapi.SerializedReference{}
+		annotation := pod.Annotations[kapi.CreatedByAnnotation]
+		err := runtime.DecodeInto(decoder, []byte(annotation), &sref)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "failed to decode %s annotation %q of pod %s", kapi.CreatedByAnnotation, annotation, pod.Name)
+			continue
+		}
+
+		dmMeta := deploymentModel.(metav1.ObjectMetaAccessor).GetObjectMeta()
+		if sref.Reference.Kind == deploymentModel.GetObjectKind().GroupVersionKind().Kind &&
+			sref.Reference.Namespace == dmMeta.GetNamespace() &&
+			sref.Reference.Name == dmMeta.GetName() &&
+			sref.Reference.UID == dmMeta.GetUID() {
```

Comment:
I guess this could be shortened to just kind and uid. Or is there perhaps some helper, I overlooked, for matching object based on a reference? @mfojtik